scapy/contrib/automotive/doip.py

```diff
@@ -0,0 +1,241 @@
+        return 0
+
+    def hashret(self):
+        # type: () -> bytes
+        if self.payload_type in [0x8001, 0x8002, 0x8003]:
+            return bytes(self)[:2] + struct.pack(
+                "H", self.target_address ^ self.source_address)
+        return bytes(self)[:2]
+
+    def post_build(self, pkt, pay):
+        # type: (bytes, bytes) -> bytes
+        """
+        This will set the Field 'payload_length' to the correct value.
+        """
+        if self.payload_length is None:
+            pkt = pkt[:4] + struct.pack("!I", len(pay) + len(pkt) - 8) + \
+                pkt[8:]
+        return pkt + pay
+
+    def extract_padding(self, s):
+        # type: (bytes) -> Tuple[bytes, Optional[bytes]]
+        if self.payload_type == 0x8001:
+            return s[:self.payload_length - 4], None
+        else:
+            return b"", None
+
+
+class DoIPSocket(StreamSocket):
+    def __init__(self, ip='127.0.0.1', port=13400, activate_routing=True,
+                 source_address=0xe80, target_address=0,
+                 activation_type=0):
+        # type: (str, int, bool, int, int, int) -> None
+        self.ip = ip
+        self.port = port
+        self.source_address = source_address
+        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        s.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
+        s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+        s.connect((self.ip, self.port))
+        StreamSocket.__init__(self, s, DoIP)
+
+        if activate_routing:
+            resp = self.sr1(
+                DoIP(payload_type=0x5, activation_type=activation_type,
+                     source_address=source_address),
+                verbose=False, timeout=1)
+            if resp and resp.payload_type == 0x6 and \
+                    resp.routing_activation_response == 0x10:
+                self.target_address = target_address or \
+                    resp.logical_address_doip_entity
+                print("Routing activation successful! "
+                      "Target address set to: 0x%x" % self.target_address)
+            else:
+                print("Routing activation failed! Response: %s" % repr(resp))
+
+
+class UDS_DoIPSocket(DoIPSocket):
+    def send(self, x):
+        # type: (Union[Packet, bytes]) -> int
+        if isinstance(x, UDS):
+            pkt = DoIP(payload_type=0x8001, source_address=self.source_address,
+                       target_address=self.target_address) / x
+        else:
+            pkt = x
+
+        try:
+            x.sent_time = time.time()  # type: ignore
+        except AttributeError:
+            pass
+
+        return super(UDS_DoIPSocket, self).send(pkt)
+
+    def recv(self, x=MTU):
+        # type: (int) -> Packet
+        pkt = super(UDS_DoIPSocket, self).recv(x)
+        if pkt.payload_type == 0x8001:
+            return pkt[1]
+        else:
+            return pkt
+
+
+bind_layers(UDP, DoIP, sport=13400)
+bind_layers(UDP, DoIP, dport=13400)
```

Comment:
Remember with UDP: use `bind_top_bottom` for those then add a `bind_layers(UDP, DoIP, dport, sport)` for the build. Otherwise you'll have a remain (port 53) in one of the ports

Reply:
I hope I did this right.

Reply:
Hi, I'm really sorry I confused the names. (Those names really are terrible). I meant:
```python
bind_bottom_up(UDP, DoIP, sport=13400)
bind_bottom_up(UDP, DoIP, dport=13400)
bind_layers(UDP, DoIP, sport=13400, dport=13400)
```

Reply:
Thanks for this explanation.